### ISSUE TITLE
Impute missing values

### ISSUE THREAD

**Issue by jangop** [davidhealey/waistline#223]
A lot of items in [Open Food Facts](https://openfoodfacts.org/) (OFF) lack a number of values (see, for example, [this notebook on kaggle][1], or [this notebook on github][2]). Oftentimes this is simply because the producer does not disclose the information in question, see, for example [these thin-rolled oats](https://world.openfoodfacts.org/product/4388844221321). Although their fiber content is not explicitly given, OFF does contain the single ingredient, and [FoodData Central](https://fdc.nal.usda.gov/) (see #222, #154, #86) yields the respective value of 10.1 g per 100 g [for oats](https://fdc.nal.usda.gov/fdc-app.html#/food-details/1101825/nutrients).

Could Waistline use this to (optionally) impute missing values on the fly, where it is easy?

Furthermore, how about more involved missing value imputation for multi-ingredient products? There is a paper online on this topic ([MIGHT: Statistical Methodology for Missing-Data Imputation in Food Composition Databases by G. Ispirova et al.](https://www.mdpi.com/2076-3417/9/19/4111/pdf)) and [scikit-learn has appropriate implementations](https://scikit-learn.org/stable/modules/impute.html).

[1]: https://www.kaggle.com/daveianhickey/can-we-trust-open-food-facts
[2]: https://github.com/PhilippeCodes/Open-Food-Facts-Data-Analysis/blob/master/Open%20Food%20Facts.ipynb

(This time, I searched through closed issues as well as open ones before opening this issue. 🙈)

**Comment by davidhealey** [comment]
So you mean if an OFF item is called "rolled oats" and it is missing a value, Waistline should check USDA for the missing value of generic rolled oats?

**Comment by stephanegigandet** [comment]
@jangop Inferring nutrition facts from ingredients is a very interesting topic, in particular for fibers that are very often missing on labels. We (Open Food Facts) are also considering doing the reverse: inferring the % of each ingredient (which is often not specified in the ingredients list) based on nutrition facts. For instance if the nutrient salt is 2%, then we know it is the max bound for salt as an ingredient. (we are very interested in precise % for each ingredient in order to compute environmental impacts)

> Could Waistline use this to (optionally) impute missing values on the fly, where it is easy?

May I suggest that this is something that we (anyone interested) could try to do on OFF instead? That way it would benefit all OFF reusers, and there may be more people interested in working on it.

**Comment by davidhealey** [comment]
> May I suggest that this is something that we (anyone interested) could try to do on OFF instead? That way it would benefit all OFF reusers, and there may be more people interested in working on it.

I think that's a much better idea. It seems like something beyond the scrope of Waistline anyway.

**Comment by jangop** [comment]
> So you mean if an OFF item is called "rolled oats" and it is missing a value, Waistline should check USDA for the missing value of generic rolled oats?

Almost. For a simple case, where an item contains only a single ingredient, known to OFF, Waistline could look up that ingredient on USDA to fill in missing values; the name given to the item by the brand/producer would not go into this, so as not to start guessing based on strings.

**Comment by jangop** [comment]
> May I suggest that this is something that we (anyone interested) could try to do on OFF instead? That way it would benefit all OFF reusers, and there may be more people interested in working on it.

Yes, that would be preferable; as long as it does not pollute the core database in any way. Exact and verifiable information should not suffer from imputation, which would (I suppose) be application-specific. That is why I went to the Waistline Issues first. 😉

I'll see what I can do -- and hopefully get back to @davidhealey with something usable.
